source/linux/io_epoll_context.cpp

```diff
@@ -0,0 +1,485 @@
+      throw std::system_error{errorCode, std::system_category()};
+    }
+  }
+
+  LOG("io_epoll_context construction done");
+}
+
+io_epoll_context::~io_epoll_context() {
+  epoll_event event;
+  (void)epoll_ctl(epollFd_.get(), EPOLL_CTL_DEL, remoteQueueEventFd_.get(), &event);
+  (void)epoll_ctl(epollFd_.get(), EPOLL_CTL_DEL, timerFd_.get(), &event);
+  LOG("io_epoll_context destructor done");
+}
+
+void io_epoll_context::run_impl(const bool& shouldStop) {
+  LOG("run loop started");
+
+  auto* oldContext = std::exchange(currentThreadContext, this);
+  scope_guard g = [=]() noexcept {
+    std::exchange(currentThreadContext, oldContext);
+    LOG("run loop exited");
+  };
+
+  while (true) {
+    // Dequeue and process local queue items (ready to run)
+    execute_pending_local();
+
+    if (shouldStop) {
+      break;
+    }
+
+    // Check for any new completion-queue items.
+    acquire_completion_queue_items();
+
+    if (timersAreDirty_) {
+      update_timers();
+    }
+
+    // Check for remotely-queued items.
+    // Only do this if we haven't submitted a poll operation for the
+    // completion queue - in which case we'll just wait until we receive the
+    // completion-queue item).
+    if (!remoteQueueReadSubmitted_) {
+      acquire_remote_queued_items();
+    }
+
+    if (localQueue_.empty()) {
+      if (!remoteQueueReadSubmitted_) {
+        LOG("try_register_remote_queue_notification()");
+        remoteQueueReadSubmitted_ = try_register_remote_queue_notification();
+      }
+
+      LOG("epoll_wait()");
+      int result = epoll_wait(
+        epollFd_.get(),
+        completed_,
+        maxCount_,
+        -1);
+      if (result < 0) {
+        int errorCode = errno;
+        throw std::system_error{errorCode, std::system_category()};
+      }
+      LOGX("epoll_wait() returned %i\n", result);
+      count_ = result;
+    }
+  }
+}
+
+bool io_epoll_context::is_running_on_io_thread() const noexcept {
+  return this == currentThreadContext;
+}
+
+void io_epoll_context::schedule_impl(operation_base* op) {
+  assert(op != nullptr);
+  if (is_running_on_io_thread()) {
+    LOG("schedule_impl - local");
+    schedule_local(op);
+  } else {
+    LOG("schedule_impl - remote");
+    schedule_remote(op);
+  }
+}
+
+void io_epoll_context::schedule_local(operation_base* op) noexcept {
+  LOG("schedule_local");
+  localQueue_.push_back(op);
+}
+
+void io_epoll_context::schedule_local(operation_queue ops) noexcept {
+  localQueue_.append(std::move(ops));
+}
+
+void io_epoll_context::schedule_remote(operation_base* op) noexcept {
+  LOG("schedule_remote");
+  bool ioThreadWasInactive = remoteQueue_.enqueue(op);
+  if (ioThreadWasInactive) {
+    // We were the first to queue an item and the I/O thread is not
+    // going to check the queue until we signal it that new items
+    // have been enqueued remotely by writing to the eventfd.
+    signal_remote_queue();
```

Comment:
This method is potentially-throwing (if signalling the eventfd fails for some reason).
Should we remove the noexcept declaration here?

Reply:
I replaced the throw with an explicit terminate.